p4-16/spec/P4-16-spec.mdk

```diff
@@ -2208,16 +2208,26 @@ structField
 ~ End P4Grammar
 
 where each `typeRef` is restricted to a bit-string type (fixed or
-variable) or an integer type. This declaration introduces a new
-identifier in the current scope; the type can be referred to using
-this identifier. A header is similar to a `struct` in C,
+variable), a fixed-width signed integer type, a boolean type, or a struct that itself contains other 
+struct fields, nested arbitrarily, as long as all of the "leaf" types are
+are `bit<W>`, `int<W>`, a Serializable Enum, or a `bool`.  If a `bool` is used
+inside a P4 header, all implementations encode the `bool` as a one bit long field. 
+This declaration introduces a new identifier in the current scope; the 
+type can be referred to using this identifier. A header is similar to a `struct` in C,
 containing all the specified fields. However, in addition, a header also
 contains a hidden Boolean "validity" field. When the "validity" bit is `true`
 we say that the "header is valid". When a header is created
 its "validity" bit is automatically set to `false`. The "validity"
 bit can be manipulated by using the header methods `isValid()`, `setValid()`,
 and `setInvalid()`, as described in Section [#sec-ops-on-hdrs].
 
+Note, nesting of headers is not supported.  The reason is if one used an 
+extract() on a header that contained nested headers, then all of the headers 
+involved would become valid.  Additionally, what about invoking `hdr.setInvalid()` on a 
+"sub-header", without doing so on the top level header? What if one tried to then emit() 
+the top level header in that state? What should be emitted?  Due to such reasons, nested
+headers are not supported.
```

Comment:
Sorry for so many comments here, but wanted to suggest some alternate phrasing for this paragraph.  I do not think there is any technically wrong with it -- it is simply worded in a way that sounds a bit more like giving a talk, or a conversation, versus somewhat more dry formal technical presentation that one typically sees in language specifications :-)  A suggested alternative is given below:

"Nesting of headers is not supported.  One reason is that headers containing structs covers many of the use cases where headers containing other headers might be useful.  Another is that it leads to complications in defining the behavior of arbitrary sequences of `setValid`, `setInvalid`, and `emit` operations.  Consider an example where header `h1` contains header `h2` as a member, both currently valid.  A program executes `h2.setInvalid()` followed by `packet.emit(h1)`.  Should all fields of `h1` be emitted, but skipping `h2`?  Similarly, should `h1.setInvalid()` invalidate all headers contained within `h1`, regardless of how deeply they are nested?"

Reply:
Thanks for the review - great feedback!  I will take your proposed text and update the doc.